### PR TITLE
[bgpcfg]: Batch bgp updates

### PR DESCRIPTION
**Pull request by pavel-shirshov**
**- Why I did it**
vtysh -f command is slow. It sometimes takes about 3 seconds.
When we need to run many vtysh -f commands that slows down the system.
With this fix bgcfgd configuration time reduced from 10.58 sec to 2.80 sec on my dut.

**- How I did it**
1. Read as many messages as possible from the database
2. Run through bgpcfgd all the messages, collecting  frr updates;
3. commit all updates as a batch

**- How to verify it**
Build an image and run on your dut. All tests must be working

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [x] 201911
- [x] 202006

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by lguohan** [comment]
@qiluo-msft , can you review this? @pavel-shirshov , do you have the performance number before and after?

**Comment by pavel-shirshov** [comment]
@lguohan I added performance change to the description

File: src/sonic-bgpcfgd/bgpcfgd/frr.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        stop_time = datetime.datetime.now() + datetime.timedelta(seconds=seconds)
+        log_info("Start waiting for FRR daemons: %s" % str(datetime.datetime.now()))
+        while datetime.datetime.now() < stop_time:
+            ret_code, out, err = run_command(["vtysh", "-c", "show daemons"], hide_errors=True)
+            if ret_code == 0 and all(daemon in out for daemon in self.daemons):
+                log_info("All required daemons have connected to vtysh: %s" % str(datetime.datetime.now()))
+                return
+            else:
+                log_warn("Can't read daemon status from FRR: %s" % str(err))
+            time.sleep(0.1)  # sleep 100 ms
```
Comment (qiluo-msft) [inline src/sonic-bgpcfgd/bgpcfgd/frr.py:30]: >time.sleep(0.1)  # sleep 100 ms [](start = 12, length = 31)

What happens if no sleep?

**Review by qiluo-msft — APPROVED** [review]
runner.py LGTM. Not familiar with other parts.

File: src/sonic-bgpcfgd/bgpcfgd/frr.py (hunk 1 above)
Comment (pavel-shirshov) [reply src/sonic-bgpcfgd/bgpcfgd/frr.py:30]: Higher cpu usage. I think it is better to wait for 0.1 than to burn cpu by requests.

File: src/sonic-bgpcfgd/tests/test_ipv6_nexthop_global.py [hunk 2]
```diff
@@ -90,14 +90,13 @@ def check_routemap_in_file(filename, route_map_name):
             found_entry = False
             found_seq_no = None
         if route_map_re.match(line):
-            found_seq_no = None
```
Comment (lguohan) [inline src/sonic-bgpcfgd/tests/test_ipv6_nexthop_global.py:93]: this should go to the other pr.

File: src/sonic-bgpcfgd/bgpcfgd/config.py [hunk 3]
```diff
@@ -1,72 +1,66 @@
...
-            err_tuple = str(cmd), ret_code, out, err
-            log_err("ConfigMgr::push(): can't push configuration '%s', rc='%d', stdout='%s', stderr='%s'" % err_tuple)
-        if ret_code == 0:
-            self.current_config = None  # invalidate config
-            self.current_config_raw = None
-        return ret_code == 0
+        if self.changes.strip() == "":
+            return True
+        rc_write = self.frr.write(self.changes)
+        rc_restart = self.frr.restart_peer_groups(self.peer_groups_to_restart)
```
Comment (lguohan) [inline src/sonic-bgpcfgd/bgpcfgd/config.py:61]: i am not sure if it is always correct to restart peer group after we make any change. for example, if we shutdown a bgp session, which peer group to restart?

File: src/sonic-bgpcfgd/bgpcfgd/frr.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    def write(config_text):
+        fd, tmp_filename = tempfile.mkstemp(dir='/tmp')
+        os.close(fd)
+        with open(tmp_filename, 'w') as fp:
+            fp.write("%s\n" % config_text)
+        command = ["vtysh", "-f", tmp_filename]
+        ret_code, out, err = run_command(command)
+        if ret_code != 0:
+            err_tuple = tmp_filename, ret_code, out, err
+            log_err("ConfigMgr::commit(): can't push configuration from file='%s', rc='%d', stdout='%s', stderr='%s'" % err_tuple)
```
Comment (lguohan) [inline src/sonic-bgpcfgd/bgpcfgd/frr.py:51]: do we want to keep this tmp file, or just put the tmpfile content into syslog?

**Comment by lguohan** [comment]
if we are doing this config bgp shutdown all, which managers are we triggering? I am not able to identify.

File: src/sonic-bgpcfgd/bgpcfgd/config.py (hunk 3 above)
Comment (pavel-shirshov) [reply src/sonic-bgpcfgd/bgpcfgd/config.py:61]: self.peer_groups_to_restart would be empty in 'shutdown a bgp session" situation and restart_peer_groups will do nothing in this case. I update self.peer_groups_to_restart in two cases now: BBR and allow_prefix.

File: src/sonic-bgpcfgd/bgpcfgd/frr.py (hunk 4 above)
Comment (pavel-shirshov) [reply src/sonic-bgpcfgd/bgpcfgd/frr.py:51]: The tmp file content it is full config on the start. Kind of a lot. Do we want to see it in the syslog?

**Comment by pavel-shirshov** [comment]
@lguohan 
managers_bgp.py
It subscribed on BGP peer object change
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-bgpcfgd/bgpcfgd/main.py#L46
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-bgpcfgd/bgpcfgd/managers_bgp.py#L153

**Comment by pavel-shirshov** [comment]
retest vs please

**Review by lguohan — APPROVED** [review]

**Comment by pavel-shirshov** [comment]
retest mellanox please

**Comment by pavel-shirshov** [comment]
retest  vsimage please

**Comment by pavel-shirshov** [comment]
retest vsimage please

**Pull request merged**